src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs

```diff
@@ -0,0 +1,312 @@
+                {
+                    if (!CompareTypeLayout(type1, type2))
+                    {
+                        return false;
+                    }
+                }
+                return true;
+
+                static bool CompareTypeLayout(MetadataType type1, MetadataType type2)
+                {
+                    // Types must either be Sequential or Explicit layout
+                    if (type1.IsSequentialLayout != type2.IsSequentialLayout)
+                    {
+                        return false;
+                    }
+
+                    if (type1.IsExplicitLayout != type2.IsExplicitLayout)
+                    {
+                        return false;
+                    }
+
+                    if (!(type1.IsSequentialLayout || type1.IsExplicitLayout))
+                    {
+                        return false;
+                    }
+
+                    bool explicitLayout = type1.IsExplicitLayout;
+
+                    // they must have the same charset
+                    if (type1.PInvokeStringFormat != type2.PInvokeStringFormat)
+                    {
+                        return false;
+                    }
+
+                    var layoutMetadata1 = type1.GetClassLayout();
+                    var layoutMetadata2 = type2.GetClassLayout();
+                    if ((layoutMetadata1.PackingSize != layoutMetadata2.PackingSize) ||
+                        (layoutMetadata1.Size != layoutMetadata2.Size))
+                        return false;
+
+                    if ((explicitLayout) && !(layoutMetadata1.Offsets == null && layoutMetadata2.Offsets == null))
+                    {
+                        if (layoutMetadata1.Offsets == null)
+                            return false;
+
+                        if (layoutMetadata2.Offsets == null)
+                            return false;
+
+                        for (int index = 0; index < layoutMetadata1.Offsets.Length; index++)
+                        {
+                            if (layoutMetadata1.Offsets[index].Offset != layoutMetadata2.Offsets[index].Offset)
+                                return false;
+                        }
+                    }
+
+                    return true;
+                }
+
+                static FieldDesc GetNextTypeEquivalentField(IEnumerator<FieldDesc> fieldEnum, bool enumMode, out bool fieldNotValidInEquivalentTypeFound)
+                {
+                    fieldNotValidInEquivalentTypeFound = false;
+                    while (fieldEnum.MoveNext())
+                    {
+                        var field = fieldEnum.Current;
+
+                        if (field.GetEffectiveVisibility() == EffectiveVisibility.Public && !field.IsStatic)
+                            return field;
+
+                        // Only public instance fields, and literal fields on enums are permitted in type equivalent structures
+                        if (!enumMode || field.IsLiteral)
```

Comment:
I may be missing something but shouldn't this check be negated? I would assume that <code>!field.IsLiteral</code> is what signifies that the type doesn't support equivalence.

Reply:
Yes. Good catch, that inspired me to build a whole test suite, so I found some more issues. Sigh. Thank you though.